rust/agents/relayer/src/msg/metadata_builder.rs

```diff
@@ -137,34 +137,66 @@ impl MetadataBuilder {
         &self,
         validators: &[H256],
     ) -> eyre::Result<MultisigCheckpointSyncer> {
-        let mut checkpoint_syncers: HashMap<H160, Arc<dyn CheckpointSyncer>> = HashMap::new();
         let storage_locations = self
             .validator_announce
             .get_announced_storage_locations(validators)
             .await?;
+
         // Only use the most recently announced location for now.
-        for (i, validator_storage_locations) in storage_locations.iter().enumerate() {
+        let mut checkpoint_syncers: HashMap<H160, Arc<dyn CheckpointSyncer>> = HashMap::new();
+        for (&validator, validator_storage_locations) in validators.iter().zip(storage_locations) {
+            debug!(
+                ?validator,
+                ?validator_storage_locations,
+                "Attempting to load checkpoint syncer for validator"
+            );
+
             for storage_location in validator_storage_locations.iter().rev() {
-                if let Ok(conf) = CheckpointSyncerConf::from_str(storage_location) {
-                    // If this is a LocalStorage based checkpoint syncer and it's not
-                    // allowed, ignore it
-                    if matches!(conf, CheckpointSyncerConf::LocalStorage { .. })
-                        && !self.allow_local_checkpoint_syncers
-                    {
-                        trace!(
-                            ?conf,
-                            "Ignoring disallowed LocalStorage based checkpoint syncer"
-                        );
-                        continue;
-                    }
+                let Ok(config) = CheckpointSyncerConf::from_str(storage_location) else {
+                    debug!(?validator, ?storage_location, "Could not parse checkpoint syncer config for validator");
+                    continue
+                };
+
+                // If this is a LocalStorage based checkpoint syncer and it's not
+                // allowed, ignore it
+                if !self.allow_local_checkpoint_syncers
+                    && matches!(config, CheckpointSyncerConf::LocalStorage { .. })
+                {
+                    debug!(
+                        ?config,
+                        "Ignoring disallowed LocalStorage based checkpoint syncer"
+                    );
+                    continue;
+                }
 
-                    if let Ok(checkpoint_syncer) = conf.build(None) {
-                        checkpoint_syncers
-                            .insert(H160::from(validators[i]), checkpoint_syncer.into());
+                match config.build(None) {
+                    Ok(checkpoint_syncer) => {
+                        // found the syncer for this validator
+                        debug!(
```

Comment:
Again ok to show only in case of failure

Reply:
Done